kafka-rest/src/main/java/io/confluent/kafkarest/response/StreamingResponse.java

```diff
@@ -141,31 +154,56 @@ private ResultOrError handleNext(T result, @Nullable Throwable error) {
   private static class InputStreamingResponse<T> extends StreamingResponse<T> {
 
     private final MappingIterator<T> mappingIteratorInput;
+    private final StreamingResponseIdleTimeLimiter idleTimeLimiter;
+
+    private Throwable hasNextError = null;
+    private boolean closed = false;
 
     private InputStreamingResponse(
-        MappingIterator<T> mappingIteratorInput, ChunkedOutputFactory chunkedOutputFactory) {
+        MappingIterator<T> mappingIteratorInput,
+        ChunkedOutputFactory chunkedOutputFactory,
+        StreamingResponseIdleTimeLimiter idleTimeLimiter) {
       super(chunkedOutputFactory);
       this.mappingIteratorInput = requireNonNull(mappingIteratorInput);
+      this.idleTimeLimiter = requireNonNull(idleTimeLimiter);
     }
 
     public void close() {
-      try {
-        mappingIteratorInput.close();
-      } catch (IOException e) {
-        log.error("Error when closing the request stream.", e);
-      }
+      closed = true;
     }
 
     @Override
     public boolean hasNext() {
-      return mappingIteratorInput.hasNext();
+      if (closed) {
+        return false;
+      }
+      try {
+        return idleTimeLimiter.callWithTimeout(mappingIteratorInput::hasNextValue);
+      } catch (ExecutionException | UncheckedExecutionException e) {
+        hasNextError = e.getCause();
+        return true;
+      } catch (InterruptedException | MaxIdleTimeExceededException e) {
+        hasNextError = e;
+        return true;
```

Comment:
Is returning `true` here meant to always delegate to `next()` to actually close the stream? Arguably in both cases here (the Jetty handler thread being interrupted for some reason, or our idle time exception being raised), we need to unconditionally close the stream.

Reply:
Yes. We delegate the responsibility to `next()` so that we have the chance to push a last message to the user saying "timeout, get lost". Notice that `next()` will `close()` the stream immediately when it sees this error, and the next `hasNext` will return false.